File: summer/WebContent/js/MS/utility/DTypeMap.js
/**
 * Second check 2013-12-14
 * DTypeMap
 */

define(["dojo/_base/declare", "system/Type"], function(declare, Type){
	var DTypeMap = declare("DTypeMap", null,{
		constructor:function(/*int*/ entryCount) 
        {
            // Constant Time Lookup entries (array size) 
            this._entryCount = entryCount; 
            this._entries = []; //new object[_entryCount];
            this._entries.length = this._entryCount;
            for(var i=0; i<entryCount; i++){
            	this._entries[i] = null;
            }
            
            this._activeDTypes = new ItemStructList/*<DependencyObjectType>*/(128);
            this._overFlow = null;
		},
        
        Get:function(dType)
        { 
            if (dType.Id < this._entryCount) 
            {
                return this._entries[dType.Id]; 
            }
            else
            {
                if (this._overFlow != null) 
                {
                    return this._overFlow.Get(dType); 
                } 

                return null; 
            }
        },

        Set:function(dType, value) 
        {
            if (dType.Id < this._entryCount) 
            { 
            	this._entries[dType.Id] = value;
            } 
            else
            {
                if (this._overFlow == null)
                { 
                	this._overFlow = new Hashtable();
                } 

                this._overFlow.Set(dType, value);
            } 

            this._activeDTypes.Add(dType);
        },
        
        // Clear the data-structures to be able to start over
//        public void 
        Clear:function() 
        {
            for (var i=0; i<this._entryCount; i++) 
            { 
            	this._entries[i] = null;
            } 

            for (var i=0; i<this._activeDTypes.Count; i++)
            {
            	this._activeDTypes.List[i] = null; 
            }
 
            if (this._overFlow != null) 
            {
            	this._overFlow.Clear(); 
            }
        }
	});
	
	Object.defineProperties(DTypeMap.prototype,{
	    // Return list of non-null DType mappings 
	    ActiveDTypes: 
	    {
	        get:function() { return this._activeDTypes; } 
	    }
	});
	
	DTypeMap.Type = new Type("DTypeMap", DTypeMap, [Object.Type]);
	return DTypeMap;
});

File: org.summer.view.widget/src/org/summer/view/widget/utils/DTypeMap.java
package org.summer.view.widget.utils;

import org.summer.view.widget.DependencyObjectType;
import org.summer.view.widget.ItemStructList;
import org.summer.view.widget.collection.Hashtable;

//[FriendAccessAllowed] // Built into Core, also used by Framework. 
/*internal*/ public class DTypeMap
{

    public DTypeMap(int entryCount) 
    {
        // Constant Time Lookup entries (array size) 
        _entryCount = entryCount; 
        _entries = new Object[_entryCount];
        _activeDTypes = new ItemStructList<DependencyObjectType>(128); 
    }

    public Object this[DependencyObjectType dType]
    { 
        get
        { 
            if (dType.Id < _entryCount) 
            {
                return _entries[dType.Id]; 
            }
            else
            {
                if (_overFlow != null) 
                {
                    return _overFlow[dType]; 
                } 

                return null; 
            }
        }

        set 
        {
            if (dType.Id < _entryCount) 
            { 
                _entries[dType.Id] = value;
            } 
            else
            {
                if (_overFlow == null)
                { 
                    _overFlow = new Hashtable();
                } 

                _overFlow[dType] = value;
            } 

            _activeDTypes.Add(dType);
        }
    } 

    // Return list of non-null DType mappings 
    public ItemStructList<DependencyObjectType> ActiveDTypes 
    {
        get { return _activeDTypes; } 
    }

    // Clear the data-structures to be able to start over
    public void Clear() 
    {
        for (int i=0; i<_entryCount; i++) 
        { 
            _entries[i] = null;
        } 

        for (int i=0; i<_activeDTypes.Count; i++)
        {
            _activeDTypes.List[i] = null; 
        }

        if (_overFlow != null) 
        {
            _overFlow.Clear(); 
        }
    }

    private int _entryCount; 
    private Object[] _entries;
    private Hashtable _overFlow; 
    private ItemStructList<DependencyObjectType> _activeDTypes; 
}